### PR TITLE
feat: delete resources when the destination is deleted

### LINKED ISSUE #232: feat: cascade delete resources when the destination is deleted
Currently, when a destination is deleted, the documents scheduled to that destination won't be removed from the state store. Furthermore, the `work` and the `workplacement` for any requested resource will remain in the platform cluster as well (although marked as `mischeduled`).

That means there's no way for customers to have Kratix cleaning up the State Store once a Destination is decommissioned. To get their state store back to where it was prior to registering it against the platform cluster, users need to:
* delete all promises and resources that got scheduled to that destination
* delete the test resources kratix created on registration

_the cleaning up of the test resources is being tracked in #231_

> [!NOTE]
Whether deleting a destination _should_ clean up the state store is up for debate. We could maintain the existing behaviour by default, and introduce a `spec` field to control the deletion, instead of just deleting it all by default. The scenarios below are assuming that that's what we want.

## Scenarios

```gherkin
# current behaviour by default
Given a destination registered with no cleanup option
When I delete the destination
Then no documents are removed from the state store

# current behaviour explicitly set
Given a destination registered with cleanup set to none
When I delete the destination
Then no documents are removed from the state store

# new cleanup behaviour
Given a destination registered with cleanup set to all
When I delete the destination
Then all workplacements for that destination are also removed
And all documents are removed from the state store
```

### PR DESCRIPTION
**Pull request by alparslanavci**
Closes #232.

Introducing the `cleanup` field in the `Destination` spec.  If set to `all`, `workplacements` and State Store contents will be removed after removing the destination. Defaults to `none`.

### CODE REVIEW

File: controllers/destination_controller.go [hunk 1]
```diff
@@ -68,6 +73,11 @@ func (r *DestinationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		logger: logger,
 	}
 
+	if destination.GetCleanup() == v1alpha1.DestinationCleanupAll &&
+		!controllerutil.ContainsFinalizer(destination, destinationCleanupFinalizer) {
+		return addFinalizers(opts, destination, []string{destinationCleanupFinalizer})
+	}
+
```
Comment (aclevername) [inline controllers/destination_controller.go:80]: Do we intend for this field to be mutable? If so we might have to support going from `all` -> `none`, which would involve removing the previously added finalizer

**Review by aclevername — COMMENTED** [review]
Thanks for PR @alparslanavci! My colleague @abangser is giving it a run locally to verify its all working 😄

File: controllers/destination_controller.go (hunk 1 above)
Comment (alparslanavci) [reply controllers/destination_controller.go]: Yes, you are right. Added the support 👍 Thanks!

**Comment by alparslanavci** [comment]
Thanks for reviewing it @aclevername! I updated it along with your comment. 👍

**Review by abangser — APPROVED** [review]
Thanks so much for the PR, excited to have this feature in! I have done the functional testing and all looks good 🙌

<details>
  <summary>Test data</summary>
  
<details>
  <summary>Default still works when field is not set ✅</summary>
  
```
$ k get destination worker-1 -o yaml | yq .spec
cleanup: none
filepath:
mode: nestedByMetadata
stateStoreRef:
kind: BucketStateStore
name: default

$ k get workplacements -A
NAMESPACE   NAME                                     AGE
default     namespace-example-793d5.worker-1-5058f   4s

$ k delete destination worker-1
[destination.platform.kratix.io](http://destination.platform.kratix.io/) "worker-1" deleted

$ k get workplacements -A
NAMESPACE   NAME                                     AGE
default     namespace-example-793d5.worker-1-5058f   25s

$ k get workplacements namespace-example-793d5.worker-1-5058f -o yaml
apiVersion: [platform.kratix.io/v1alpha1](http://platform.kratix.io/v1alpha1)
kind: WorkPlacement
metadata:
creationTimestamp: "2024-12-09T12:39:05Z"
finalizers:

- [finalizers.workplacement.kratix.io/repo-cleanup](http://finalizers.workplacement.kratix.io/repo-cleanup)
generation: 1
labels:
[kratix.io/misscheduled:](http://kratix.io/misscheduled:) "true"
[kratix.io/pipeline-name:](http://kratix.io/pipeline-name:) instance-configure
[kratix.io/targetDestinationName:](http://kratix.io/targetDestinationName:) worker-1
[kratix.io/work:](http://kratix.io/work:) namespace-example-793d5
[kratix.io/workload-group-id:](http://kratix.io/workload-group-id:) 5058f1af8388633f609cadb75a75dc9d
name: namespace-example-793d5.worker-1-5058f
namespace: default
ownerReferences:
- apiVersion: [platform.kratix.io/v1alpha1](http://platform.kratix.io/v1alpha1)
blockOwnerDeletion: true
controller: true
kind: Work
name: namespace-example-793d5
uid: 0a3bea72-6a7f-40f9-98e5-1219a7f5832e
resourceVersion: "272814"
uid: e9d23ea6-29a1-43eb-86d2-4660409a4003
spec:
id: 5058f1af8388633f609cadb75a75dc9d
promiseName: namespace
resourceName: example
targetDestinationName: worker-1
workloads:
- content: H4sIAAAAAAAA/0osyAxLLSrOzM+zUigz5MrOzEuxUvBLzE0tLkhMTuXKTS1JTEksSbTiUlDIS8xNtVIoKMrPzSxOTdHNgysCBAAA//8TUcqmRAAAAA==
filepath: namespace.yaml
status:
conditions:
- lastTransitionTime: "2024-12-09T12:39:22Z"
message: Target destination no longer matches destinationSelectors
reason: DestinationSelectorMismatch
status: "True"
type: Misscheduled
versionID: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

</details>

<details>
  <summary>Explicitly setting none works ✅</summary>
  
```
$ cat << EOF | kubectl apply -f -
pipe heredoc> apiVersion: platform.kratix.io/v1alpha1
kind: Destination
metadata:
  name: worker-1
  labels:
    environment: dev
spec:
  cleanup: none
  path: dev
  stateStoreRef:
    name: default
    kind: BucketStateStore
pipe heredoc> EOF
destination.platform.kratix.io/worker-1 created

$ k get destination worker-1 -o yaml | yq .spec
cleanup: none
filepath:
  mode: nestedByMetadata
path: dev
stateStoreRef:
  kind: BucketStateStore
  name: default

$ kaf https://raw.githubusercontent.com/syntasso/kratix-marketplace/refs/heads/main/namespace/resource-request.yaml
namespace.marketplace.kratix.io/example created
[12:51:01] (⎈|kind-platform:N/A) [~/dev/syntasso/kratix] git:(add-destination-cleanup ✔)
$ k get workplacement
NAME                                     AGE
namespace-example-02301.worker-1-5058f   3s

$ k delete destination worker-1
destination.platform.kratix.io "worker-1" deleted

$ k get workplacement -o yaml
apiVersion: v1
items:
- apiVersion: platform.kratix.io/v1alpha1
  kind: WorkPlacement
  metadata:
    creationTimestamp: "2024-12-09T12:51:03Z"
    finalizers:
    - finalizers.workplacement.kratix.io/repo-cleanup
    generation: 1
    labels:
      kratix.io/misscheduled: "true"
      kratix.io/pipeline-name: instance-configure
      kratix.io/targetDestinationName: worker-1
      kratix.io/work: namespace-example-02301
      kratix.io/workload-group-id: 5058f1af8388633f609cadb75a75dc9d
    name: namespace-example-02301.worker-1-5058f
    namespace: default
    ownerReferences:
    - apiVersion: platform.kratix.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Work
      name: namespace-example-02301
      uid: 3c63bef0-c944-40ae-8acb-2329d04376df
    resourceVersion: "274246"
    uid: 34cb31da-503d-46df-8797-62d9320f64de
  spec:
    id: 5058f1af8388633f609cadb75a75dc9d
    promiseName: namespace
    resourceName: example
    targetDestinationName: worker-1
    workloads:
    - content: H4sIAAAAAAAA/0osyAxLLSrOzM+zUigz5MrOzEuxUvBLzE0tLkhMTuXKTS1JTEksSbTiUlDIS8xNtVIoKMrPzSxOTdHNgysCBAAA//8TUcqmRAAAAA==
      filepath: namespace.yaml
  status:
    conditions:
    - lastTransitionTime: "2024-12-09T12:51:14Z"
      message: Target destination no longer matches destinationSelectors
      reason: DestinationSelectorMismatch
      status: "True"
      type: Misscheduled
    versionID: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
kind: List
metadata:
  resourceVersion: ""
```

</details>

<details>
  <summary>Set all and see it clean up ✅</summary>
  
```
$ cat << EOF | kubectl apply -f -
apiVersion: platform.kratix.io/v1alpha1
kind: Destination
metadata:
  name: worker-1
  labels:
    environment: dev
spec:
  cleanup: all
  path: dev
  stateStoreRef:
    name: default
    kind: BucketStateStore
EOF
destination.platform.kratix.io/worker-1 created

$ kaf https://raw.githubusercontent.com/syntasso/kratix-marketplace/refs/heads/main/namespace/resource-request.yaml
namespace.marketplace.kratix.io/example created

$ k get workplacement -o yaml
apiVersion: v1
items:
- apiVersion: platform.kratix.io/v1alpha1
  kind: WorkPlacement
  metadata:
    creationTimestamp: "2024-12-09T12:55:36Z"
    finalizers:
    - finalizers.workplacement.kratix.io/repo-cleanup
    generation: 1
    labels:
      kratix.io/pipeline-name: instance-configure
      kratix.io/targetDestinationName: worker-1
      kratix.io/work: namespace-example-7ff8e
      kratix.io/workload-group-id: 5058f1af8388633f609cadb75a75dc9d
    name: namespace-example-7ff8e.worker-1-5058f
    namespace: default
    ownerReferences:
    - apiVersion: platform.kratix.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Work
      name: namespace-example-7ff8e
      uid: cf55cc22-a5ea-4cca-9e6a-5ff1f32561fd
    resourceVersion: "274792"
    uid: 28999ec9-52b8-499c-9f78-0860dc5384d0
  spec:
    id: 5058f1af8388633f609cadb75a75dc9d
    promiseName: namespace
    resourceName: example
    targetDestinationName: worker-1
    workloads:
    - content: H4sIAAAAAAAA/0osyAxLLSrOzM+zUigz5MrOzEuxUvBLzE0tLkhMTuXKTS1JTEksSbTiUlDIS8xNtVIoKMrPzSxOTdHNgysCBAAA//8TUcqmRAAAAA==
      filepath: namespace.yaml
  status:
    versionID: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
kind: List
metadata:
  resourceVersion: ""

$ k delete destination worker-1
destination.platform.kratix.io "worker-1" deleted

$ k get destinations
No resources found

$ k get works namespace-example-7ff8e -o yaml
apiVersion: platform.kratix.io/v1alpha1
kind: Work
metadata:
  creationTimestamp: "2024-12-09T12:55:36Z"
  finalizers:
  - kratix.io/work-cleanup
  generation: 1
  labels:
    kratix.io/pipeline-name: instance-configure
    kratix.io/promise-name: namespace
    kratix.io/resource-name: example
    kratix.io/work-type: resource
  name: namespace-example-7ff8e
  namespace: default
  resourceVersion: "274847"
  uid: cf55cc22-a5ea-4cca-9e6a-5ff1f32561fd
spec:
  promiseName: namespace
  resourceName: example
  workloadGroups:
  - directory: .
    id: 5058f1af8388633f609cadb75a75dc9d
    workloads:
    - content: H4sIAAAAAAAA/0osyAxLLSrOzM+zUigz5MrOzEuxUvBLzE0tLkhMTuXKTS1JTEksSbTiUlDIS8xNtVIoKMrPzSxOTdHNgysCBAAA//8TUcqmRAAAAA==
      filepath: namespace.yaml
status:
  conditions:
  - lastTransitionTime: "2024-12-09T12:55:55Z"
    message: 'No Destinations available work WorkloadGroups: [5058f1af8388633f609cadb75a75dc9d]'
    reason: UnscheduledWorkloadGroups
    status: "False"
    type: Scheduled
  - lastTransitionTime: "2024-12-09T12:55:55Z"
    message: WorkGroups that have been scheduled are at the correct Destination(s)
    reason: ScheduledToCorrectDestinations
    status: "False"
    type: Misscheduled

$ k get workplacements
No resources found in default namespace.

$ mc ls -r kind/kratix
```

</details>
</details>

**Comment by alparslanavci** [comment]
Thanks @abangser, for merging it! 👍

**Comment by alparslanavci** [comment]
Also, this fixes https://github.com/syntasso/kratix/issues/231 too!

**Comment by abangser** [comment]
You're 100% right, closing that as well 🎉

**Pull request merged**